proposals/eni.md

```diff
@@ -0,0 +1,129 @@
+                                                        |
+                          +-------------------------------------------------------------+
+                          |default(global) namespace    |                               |
+                          |                             |                               |
+                          |                     +-------+-------+       +-------------+ |  +------------+
+                          |                     |    ve_br_en1  |       |             | |  |net="bridge"|
+             172.31.0.226 |   +-----------+     |               |       |             +-|--|container   |
+                       +----+ |           |     |ecs-eni bridge |       |             | |  +------------+
+                       |eth0| | ECS Agent |     |               |       | docker0     | |
+                       +----+ +-----------+     |               |       | bridge      | |
+                          |                     |               |       |             | |
+                          |                     |  169.254.x.1  |       |             | |
+                          |                     +---------------+       +-------------+ |
+                          |                                                             |
+                          +-------------------------------------------------------------+
+
+In this diagram, the primary network interface is still being used on the instance
+by the `docker0` bridge to let containers launched with `net=bridge` to communicate
+with entities outside the instance. A secondary network interface has also been
+attached for the purpose of being used by a task (`eth1`). This interface has been
+moved to the `ENI-1` network namespace. All containers of the task will share this
+network namespace.
+
+Tasks (containers in tasks) launched into these namespaces will be addressable
+by the primary IP Address of the respective secondary network interfaces. For
+example, all containers of the task using ENI `ENI-1` will be addressable within
+the VPC using the IP Address `172.31.13.14`.
+
+The `ecs-eni-br` bridge is used by the containers to communicate with the
+primary network interface `eth0` for credentials and metadata requests. The
+`veth` pairs between the `ENI-1 namespace` and the `ecs-eni-br` bridge are used
+for this purpose (`ve-en1` and `ve_br_en1`).
+
+ECS Agent will start a `pause` container for performing this configuration. This
+container will only be used to start a process that waits for a signal to
+terminate by invoking the `pause()` POSIX API. ECS Agent then moves the ENI to
+the network namespace of this container using a CNI plugin. It will then launch
+containers of the task with `--net=container:pause-container-id`, thus ensuring
+that they use the `pause` container's network namespace.
+
+#### CNI Plugin Sequence
+1. Plugin to assign ENI to a network namespace:
```

Comment:
We are planning to chain the execution from the ECS Agent, which would the responsible party for invoking these plugins. Although, this could change depending on the implementation details